qiime2/core/tests/test_validate.py

```diff
@@ -0,0 +1,279 @@
+
+        validator_object.add_validator(first_record)
+        validator_object.add_validator(second_record)
+
+        self.assertEqual(validator_object.validators,
+                         [first_record, second_record])
+        self.assertTrue(validator_object._is_sorted)
+
+    def test_run_validators(self):
+
+        validator_object = ValidationObject(IntSequence1)
+
+        has_run = False
+
+        def test_validator_method(data: list, level):
+            nonlocal has_run
+            has_run = True
+            self.assertEqual(data, [0, 1, 2])
+            self.assertEqual(level, 'max')
+
+        test_record = ValidatorRecord(validator=test_validator_method,
+                                      view=list, plugin='this_plugin',
+                                      context=IntSequence1)
+
+        validator_object.add_validator(test_record)
+
+        validator_object(self.simple_int_seq, level='max')
+
+        self.assertTrue(has_run)
+
+    def test_run_validators_validation_exception(self):
+        validator_object = ValidationObject(AscIntSequence)
+
+        def test_raising_validation_exception(data: list, level):
+            raise ValidationError("2021-08-24")
+
+        test_record = ValidatorRecord(
+                          validator=test_raising_validation_exception,
+                          view=list, plugin='this_plugin',
+                          context=AscIntSequence)
+
+        validator_object.add_validator(test_record)
+
+        with self.assertRaisesRegex(ValidationError,
+                                    "2021-08-24"):
+            validator_object(data=[], level=None)
+
+    def test_run_validators_unknown_exception(self):
+        validator_object = ValidationObject(AscIntSequence)
+
+        def test_raising_validation_exception(data: list, level):
+            raise KeyError("2021-08-24")
+
+        test_record = ValidatorRecord(
+                          validator=test_raising_validation_exception,
+                          view=list, plugin='this_plugin',
+                          context=AscIntSequence)
+
+        validator_object.add_validator(test_record)
+
+        with self.assertRaisesRegex(ImplementationError,
+                                    "attempted to validate"):
+            validator_object(data=[], level=None)
+
+    def test_validator_sorts(self):
+        self.pm = PluginManager()
+
+        test_object = self.pm.validators[Squid]
+
+        self.assertFalse(test_object._is_sorted)
+
+        exp = ['validator_sort_first',
+               'validator_sort_middle',
+               'validator_sort_middle_b',
+               'validator_sort_last']
+
+        exp2 = ['validator_sort_first',
+                'validator_sort_middle_b',
+                'validator_sort_middle',
+                'validator_sort_last']
+
+        obs = [record.validator.__name__ for record in test_object.validators]
+
+        self.assertIn(obs, [exp, exp2])
```

Comment:
nice approach to the ambiguity here! (which is worth preserving to keep the tests from being brittle to the implementation)

Reply:
Thanks!